contracts/truefi2/Liquidator2.sol

```diff
@@ -0,0 +1,109 @@
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+
+import {ILoanToken2} from "./interface/ILoanToken2.sol";
+import {IPoolFactory} from "./interface/IPoolFactory.sol";
+import {ITrueFiPool2} from "./interface/ITrueFiPool2.sol";
+import {IStakingPool} from "../truefi/interface/IStakingPool.sol";
+import {ITrueFiPoolOracle} from "./interface/ITrueFiPoolOracle.sol";
+import {ILoanFactory2} from "./interface/ILoanFactory2.sol";
+import {SafeMath} from "@openzeppelin/contracts/math/SafeMath.sol";
+
+/**
+ * @title Liquidator2
+ * @notice Liquidate LoanTokens with this Contract
+ * @dev When a Loan becomes defaulted, Liquidator allows to
+ * compensate pool participants, by transferring some of TRU to the pool
+ */
+contract Liquidator2 is UpgradeableClaimable {
+    using SafeMath for uint256;
+
+    // ================ WARNING ==================
+    // ===== THIS CONTRACT IS INITIALIZABLE ======
+    // === STORAGE VARIABLES ARE DECLARED BELOW ==
+    // REMOVAL OR REORDER OF VARIABLES WILL RESULT
+    // ========= IN STORAGE CORRUPTION ===========
+
+    IPoolFactory public poolFactory;
+    IStakingPool public stkTru;
+    IERC20 public tru;
+    ILoanFactory2 public loanFactory;
+
+    // max share of tru to be taken from staking pool during liquidation
+    // 1000 -> 10%
+    uint256 public fetchMaxShare;
+
+    // ======= STORAGE DECLARATION END ============
+
+    /**
+     * @dev Emitted fetch max share is changed
+     * @param newShare New share set
+     */
+    event FetchMaxShareChanged(uint256 newShare);
+
+    /**
+     * @dev Emitted when a loan gets liquidated
+     * @param loan Loan that has been liquidated
+     */
+    event Liquidated(ILoanToken2 loan);
+
+    /**
+     * @dev Initialize this contract
+     */
+    function initialize(
+        IPoolFactory _poolFactory,
+        IStakingPool _stkTru,
+        IERC20 _tru,
+        ILoanFactory2 _loanFactory
+    ) public initializer {
+        UpgradeableClaimable.initialize(msg.sender);
+
+        poolFactory = _poolFactory;
+        stkTru = _stkTru;
+        tru = _tru;
+        loanFactory = _loanFactory;
+        fetchMaxShare = 1000;
+    }
+
+    /**
+     * @dev Set new max fetch share
+     * @param newShare New share to be set
+     */
+    function setFetchMaxShare(uint256 newShare) external onlyOwner {
+        require(newShare > 0, "Liquidator: Share cannot be set to 0");
+        require(newShare <= 10000, "Liquidator: Share cannot be larger than 10000");
+        fetchMaxShare = newShare;
+        emit FetchMaxShareChanged(newShare);
+    }
+
+    /**
+     * @dev Liquidates a defaulted Loan, withdraws a portion of tru from staking pool
+     * then transfers tru to TrueFiPool as compensation
+     * @param loan Loan to be liquidated
+     */
+    function liquidate(ILoanToken2 loan) external {
+        require(loanFactory.isLoanToken(address(loan)), "Liquidator: Unknown loan");
+        ITrueFiPool2 pool = ITrueFiPool2(loan.pool());
```

Comment:
check here that loan.token() is on approvedTokens

Reply:
require(approvedLoans(pool.token() != 0), "Liquidator: Token not approved for default protection");